### PR TITLE
[SPARK-22821][TEST] Basic tests for WidenSetOperationTypes, BooleanEquality, StackCoercion and Division

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Test Coverage for `WidenSetOperationTypes`, `BooleanEquality`, `StackCoercion`  and `Division`, this is a Sub-tasks for [SPARK-22722](https://issues.apache.org/jira/browse/SPARK-22722).

## How was this patch tested?
N/A

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+SELECT true = cast(1 as int) FROM t;
+SELECT true = cast(1 as bigint) FROM t;
+SELECT true = cast(1 as float) FROM t;
+SELECT true = cast(1 as double) FROM t;
+SELECT true = cast(1 as decimal(10, 0)) FROM t;
+SELECT true = cast(1 as string) FROM t;
+SELECT true = cast('1' as binary) FROM t;
+SELECT true = cast('1' as boolean) FROM t;
+SELECT true = cast('2017-12-11 09:30:00.0' as timestamp) FROM t;
+SELECT true = cast('2017-12-11 09:30:00' as date) FROM t;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql]: Also please add the cases for `false = 0`

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+SELECT true = cast(1 as int) FROM t;
+SELECT true = cast(1 as bigint) FROM t;
+SELECT true = cast(1 as float) FROM t;
+SELECT true = cast(1 as double) FROM t;
+SELECT true = cast(1 as decimal(10, 0)) FROM t;
+SELECT true = cast(1 as string) FROM t;
+SELECT true = cast('1' as binary) FROM t;
+SELECT true = cast(1 as boolean) FROM t;
+SELECT true = cast('2017-12-11 09:30:00.0' as timestamp) FROM t;
+SELECT true = cast('2017-12-11 09:30:00' as date) FROM t;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql]: Could we just use one side?

For example, let us get rid of `true = cast(1 as tinyint)` and keep `cast(1 as tinyint) = true`?

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/division.sql [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+SELECT cast(1 as string) / cast(1 as float) FROM t;
+SELECT cast(1 as string) / cast(1 as double) FROM t;
+SELECT cast(1 as string) / cast(1 as decimal(10, 0)) FROM t;
+SELECT cast(1 as string) / cast(1 as string) FROM t;
+SELECT cast(1 as string) / cast('1' as binary) FROM t;
+SELECT cast(1 as string) / cast(1 as boolean) FROM t;
+SELECT cast(1 as string) / cast('2017-12-11 09:30:00.0' as timestamp) FROM t;
+SELECT cast(1 as string) / cast('2017-12-11 09:30:00' as date) FROM t;
+
+SELECT cast(1 as string) / cast(1 as tinyint) FROM t;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/division.sql]: Duplicates?

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/stackCoercion.sql [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+--   Unless required by applicable law or agreed to in writing, software
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+--
+
+CREATE TEMPORARY VIEW t AS SELECT 1;
+
+SELECT stack(cast(1 as tinyint), cast(1 as tinyint), null) FROM t;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/stackCoercion.sql]: This suite is not needed.

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/widenSetOperationTypes.sql [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast(2 as float) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast(2 as double) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast(2 as decimal(10, 0)) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast(2 as string) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast('2' as binary) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast(2 as boolean) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast('2017-12-11 09:30:00.0' as timestamp) FROM t;
+SELECT cast('2017-12-12 09:30:00' as date) FROM t UNION SELECT cast('2017-12-11 09:30:00' as date) FROM t;
+
+-- EXCEPT
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/widenSetOperationTypes.sql]: We do not need EXCEPT and INTERSECT

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql (hunk 2 above)
Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/booleanEquality.sql]: I think we should keep both, we have some below usage:

https://github.com/apache/spark/blob/6d7ebf2f9fbd043813738005a23c57a77eba6f47/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L486-L489

https://github.com/apache/spark/blob/6d7ebf2f9fbd043813738005a23c57a77eba6f47/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L134-L135

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**